design-documents/mde-project-structure.md

```diff
@@ -0,0 +1,41 @@
+# Overview
+
+There is a need to have unified approach to structuring the Git-based project while developing MDEs (Magento Developed Extensions), like Page Builder and MSI.
+
+Currently, MSI is developed in a fork of main Magento repository, while Page Builder is developed in a separate repository (like B2B).
+
+## MDE project structure for Git-based installations
+
+MDE project must be structured as Magento B2B:
+ 1. Modules must be placed under `app/code`
+ 1. Each module must have valid `composer.json` with no version specified. Package version must be generated and added to the `composer.json` during release publication
+ 1. No core modules must be committed to the MDE repository
+ 1. Unit and MFTF tests must be distributed between modules (since they are modular). All other tests must be put inside `dev/tests` like it is done in Magento Open Source repository (because these testing frameworks do not support modularity)
+
+## Git-based project linking for local development
+
+Preferred way of linking the project is using [Composer path repository](https://getcomposer.org/doc/05-repositories.md#path).
+
+1. Check out magento 2 open source edition from Git into `<project_root>/magento2ce`
+1. Check out MDE into `<project_root>/magento2ce/<mde-name>`
+1. Add path composer repository declaration to `<project_root>/magento2ce/composer.json`:
+    ```json
+    "minimum-stability": "dev",
+    "repositories": [
+        {
+            "type": "path",
+            "url": "./<mde-name>/app/code/*/*"
+        }
+    ]
+    ```
+1. Add dependencies on all relevant MDE modules to the `require` section of `<project_root>/magento2ce/composer.json`. Use `*` as a package version
```

Comment:
MSI has more than 30 modules. Because of high MSI modularity, there is no single module in a repo that will install all other. For the release process, we resolved this with an introduction of additional composer metapackage that includes all modules (API, implementation, UI and admin modules).

We need another approach or improve this one with an introduction of standardized composer metapackage structure or providing a tool for automatic modification of composer.json (but as mentioned in other comment, I think that we need to avoid this for git-based installations)